osu.Game/Rulesets/Objects/SliderPath.cs

```diff
@@ -305,17 +319,63 @@ private List<Vector2> calculateSubPath(ReadOnlySpan<Vector2> subControlPoints, P
                         break;
 
                     return subPath;
+                }
 
                 case SplineType.Catmull:
-                    return PathApproximator.CatmullToPiecewiseLinear(subControlPoints);
+                {
+                    List<Vector2> subPath = PathApproximator.CatmullToPiecewiseLinear(subControlPoints);
+
+                    // At draw time, osu!stable optimises paths by only keeping piecewise segments that are 6px apart.
+                    // For the most part we don't care about this optimisation, and its additional heuristics are hard to reproduce in every implementation.
+                    //
+                    // However, it matters for Catmull paths which form "bulbs" around sequential knots with identical positions,
+                    // so we'll apply a very basic form of the optimisation here and return a length representing the optimised portion.
+                    // The returned length is important so that the optimisation doesn't cause the path to get extended to match the value of ExpectedDistance.
```

Comment:
Note that "At draw time" implies that the path normally doesn't have this optimisation applied. This actually has an effect.

In osu!stable the ball behaves like this:

https://github.com/ppy/osu/assets/1329837/bea6d348-54ba-44f4-aa13-1bff4352a5b5

Notice how it does the weird loop at the knot, where there is _technically_ still a bulb there. In osu!lazer it behaves like this:

https://github.com/ppy/osu/assets/1329837/5f6af545-b49b-4b66-8a88-1350e388da82

Reply:
On a visual pass, it looks like stable is taking longer at the "bulb" points while lazer isn't. So before I dive into the code, can you explain where the extra time is spent in lazer? Or does the slider just get marginally shorter?

Reply:
It looks like the extra time is spent at the end of the slider in lazer, going by how it slows down.

Reply:
and this is what we want?

Reply:
I was going to say that this behaves better than stable, but I'm still thinking about it

Technically it's possible to mimic stable's behaviour more closely, but I want to see how the other rulesets work in diffcalc first.

Reply:
I think this is still what we want. Applying the optimisation later on (in `GetPathToProgress`) would cause it to be applied to non-Catmull segments. This is probably the best that can reasonably be done.

That being said, the diffcalc sheet showed changes in catch which I didn't like. I've changed it to only apply the optimisation in the osu ruleset.

Reply:
So what I'm getting out of this review thread is...

- Stable approximates the path in the same way that lazer does (in a way that affects _all segment types_, but catmull particularly as it's prone to these "bulbs"), but then it happens to delete the "bulbs" on catmull sliders as an optimisation, but it only does so at the point of displaying visuals, so *actual* input tracking uses the "bulbed" slider path.
- lazer displays the approximated path as it was given it, i.e. with "bulbs", but the input handling is consistent with the visuals.

Is that understanding correct?

In that case I'm not thoroughly convinced that lazer needs changing here. To use an analogy, to me this is a bit like that "fake slider follow circle radius" thing wherein stable lies with its visuals as to how large the follow circle really is.

Classic mod toggle, maybe...?

Reply:
> but it only does so at the point of displaying visuals, so actual input tracking uses the "bulbed" slider path.

Correct.

> but catmull particularly as it's prone to these "bulbs"), but then it happens to delete the "bulbs" on catmull sliders as an optimisation

The optimisation targets every path type and only deletes the bulbs by sheer luck. I'd actually say that the optimisation is more aggressive for other path types, but I digress. You can still bend the sliders to certain degrees and have the bulbs pop out again.

The optimisation traces back to the very beginnings of osu! (Sept. 2007 // b60) likely targeted at improving performance more-so than visuals, because it was batched along with another that made sliders use AA'd paths (`a6fe6288c026a3efc485981bcf2f76b9b29e377e`):

> (+) Sliders look a LOT nicer.  Performance should be considerably higher too - although I still believe I have a bit of work to do in this area.

![image](https://github.com/ppy/osu/assets/1329837/95949581-3b13-4444-b1ea-646bf964f0ac)
(2007-08-26, credit to osekai)

> Classic mod toggle, maybe...?

I don't feel okay with maps looking weird just because they were using an old algorithm. I think for all intents and purposes this should be considered as part of the algorithm. Furthermore, as mentioned in https://github.com/ppy/osu/issues/27711, there are even recent maps using Catmull.

Reply:
So to sum up, the effect of this change is such that visuals on lazer will now match stable, but the _actual_ input handling will be slightly different (because lazer will still handle input to match what it displays, it doesn't do the dumb games stable does with showing one path and using another for input).

I think I could be fine with that, personally. I definitely don't want to play the stable dumb games, so it's either this PR or nothing as far as I'm concerned. @peppy thoughts?